### PR TITLE
add second argument (parameters) to all RV compute fucntions

### LINKED ISSUE #196: Make `compute` function always take two arguments
- [ ] In graph docs
- [ ] In tests
- [ ] In type hints

### PR DESCRIPTION
**Pull request by mscroggs**
- Added second argument to all compute functions in random variable nodes
- Added parameters input to evaluate algorithms
- Updated docs to reflect changes

Resolves #196

### CODE REVIEW

File: tests/test_algorithms/test_evaluate.py [hunk 1]
```diff
@@ -13,80 +13,100 @@
 def evaluate_test_graph() -> Graph:
     return continuous_treatment_model(
         z_len=2,
-        compute_u_x=lambda data: data["c"] + data["l"][0],
-        compute_u_y=lambda data: data["c"] + 1,
-        compute_x=lambda data: data["z"][0] + data["u_x"] - data["l"][0],
-        compute_y=lambda data: data["x"] * data["u_y"],
+        compute_u_x=lambda data, _params: data["c"],
```
Comment (mscroggs) [inline tests/test_algorithms/test_evaluate.py]:
```suggestion
        compute_u_x=lambda data, _params: data["c"] + data["l"][0],
```

File: tests/test_algorithms/test_evaluate.py [hunk 2]
```diff
@@ -13,80 +13,100 @@
...
+        compute_u_y=lambda data, _params: data["c"] + 1,
+        compute_x=lambda data, _params: data["z"][0] + data["u_x"] - data["phi_x"],
```
Comment (mscroggs) [inline tests/test_algorithms/test_evaluate.py]:
```suggestion
        compute_x=lambda data, _params: data["z"][0] + data["u_x"] - data["l"][0],
```

File: tests/test_algorithms/test_evaluate.py [hunk 3]
```diff
@@ -13,80 +13,100 @@
...
+            {},
             {"u_x": 1.0},
             id="CtsRVNode evaluation, 'given that' overrides computed value",
         ),
         pytest.param(
             "u_y",
             {"l": jnp.atleast_1d(0.0), "c": 4.0},
-            {"u_x": 4.0, "u_y": 5.0},
+            {},
+            {"u_y": 5.0},
```
Comment (mscroggs) [inline tests/test_algorithms/test_evaluate.py]:
```suggestion
            {"u_x": 4.0, "u_y": 5.0},
```

File: docs/users/graph.md [hunk 4]
```diff
@@ -170,11 +170,11 @@ Python callables defining $f_r$ and $f_m$ can also be provided.
 ```python
 from causalprog.graph.continuous_treatment import continuous_treatment_model
 
-graph = continuous_treatment_model(
-    compute_u_x=lambda values: values["c"] + 1.0,
-    compute_u_y=lambda values: values["c"] * 2,
-    compute_x=lambda values: values["z"] + values["l"] - values["u_x"],
-    compute_y=lambda values: values["x"] * values["u_y"],
+graph = example_model(
```
Comment (willGraham01) [inline docs/users/graph.md]: Looks like the merge of `main` reverted the name of this function? Should be `continuous_treatment_model` but seems to have gone back to its old name.

File: docs/users/graph.md
Comment (willGraham01) [inline docs/users/graph.md:269]: Does the `graph` variable need to be re-created again in this code block? If not, it's earlier definition likely needs to be changed to reflect the fact that it's taking parameters now?

File: src/causalprog/algorithms/evaluate.py [hunk 5]
```diff
@@ -15,6 +18,7 @@ def evaluate_down_to(
         graph: The graph that the node is contained in.
         outcome_node_label: The label of the node to evaluate down to.
         values: Values taken by nodes whose value is given
+        parameters: Parameters to pass onto compute functions
```
Comment (willGraham01) [inline src/causalprog/algorithms/evaluate.py]:
```suggestion
        parameters: Parameters to pass to compute functions
```

File: src/causalprog/algorithms/evaluate.py [hunk 6]
```diff
@@ -6,7 +6,10 @@
 
 
 def evaluate_down_to(
-    graph: Graph, outcome_node_label: str, values: dict[str, float | npt.NDArray[float]]
+    graph: Graph,
+    outcome_node_label: str,
+    values: dict[str, float | npt.NDArray[float]],
+    parameters: dict[str, float | npt.NDArray[float]],
```
Comment (willGraham01) [inline src/causalprog/algorithms/evaluate.py]:
```suggestion
    parameters: dict[str, ModelParam],
```

If you `from causalprog._types import ModelParam`

File: src/causalprog/algorithms/evaluate.py [hunk 7]
```diff
@@ -33,12 +37,17 @@ def evaluate_down_to(
...
+        )
     return computed_values
 
 
 def evaluate(
-    graph: Graph, outcome_node_label: str, values: dict[str, float | npt.NDArray[float]]
+    graph: Graph,
+    outcome_node_label: str,
+    values: dict[str, float | npt.NDArray[float]],
+    parameters: dict[str, float | npt.NDArray[float]],
```
Comment (willGraham01) [inline src/causalprog/algorithms/evaluate.py]:
```suggestion
    parameters: dict[str, ModelParam],
```

as above

File: src/causalprog/algorithms/evaluate.py [hunk 8]
```diff
@@ -47,9 +56,12 @@ def evaluate(
         graph: The graph that the node is contained in.
         outcome_node_label: The label of the node to evaluate.
         values: Values taken by nodes whose value is given
+        parameters: Parameters to pass onto compute functions
```
Comment (willGraham01) [inline src/causalprog/algorithms/evaluate.py]:
```suggestion
        parameters: Parameters to pass to compute functions
```

File: src/causalprog/graph/node/base.py [hunk 9]
```diff
@@ -120,12 +120,14 @@ def sample(
     def evaluate(
         self,
         given_values: dict[str, jax.Array],
+        parameters: dict[str, jax.Array],
```
Comment (willGraham01) [inline src/causalprog/graph/node/base.py]:
```suggestion
        parameters: ModelParam,
```

(`from causalprog._types import ModelParam` proviso here too)

File: src/causalprog/graph/node/component.py [hunk 10]
```diff
@@ -54,6 +54,7 @@ def sample(
     def evaluate(
         self,
         given_values: dict[str, jax.Array],
+        parameters: dict[str, jax.Array],
```
Comment (willGraham01) [inline src/causalprog/graph/node/component.py]:
```suggestion
        parameters: ModelParam,
```

File: src/causalprog/graph/node/constant.py [hunk 11]
```diff
@@ -45,6 +45,7 @@ def sample(
     def evaluate(
         self,
         given_values: dict[str, jax.Array],
+        parameters: dict[str, jax.Array],
```
Comment (willGraham01) [inline src/causalprog/graph/node/constant.py]:
```suggestion
        parameters: ModelParam,
```

File: src/causalprog/graph/node/data.py [hunk 12]
```diff
@@ -45,6 +45,7 @@ def sample(
     def evaluate(
         self,
         given_values: dict[str, jax.Array],
+        parameters: dict[str, jax.Array],
```
Comment (willGraham01) [inline src/causalprog/graph/node/data.py]:
```suggestion
        parameters: ModelParam,
```

File: src/causalprog/graph/node/distribution.py [hunk 13]
```diff
@@ -73,6 +73,7 @@ def sample(
     def evaluate(
         self,
         given_values: dict[str, jax.Array],
+        parameters: dict[str, jax.Array],
```
Comment (willGraham01) [inline src/causalprog/graph/node/distribution.py]:
```suggestion
        parameters: ModelParam,
```

File: src/causalprog/graph/node/random_variables.py [hunk 14]
```diff
@@ -36,7 +36,8 @@ def __init__(
         *,
         shape: tuple[int, ...] = (),
         label: str,
-        compute: Callable | None = None,
+        compute: Callable[[dict[str, jax.Array], dict[str, jax.Array]], jax.Array]
+        | None = None,
```
Comment (willGraham01) [inline src/causalprog/graph/node/random_variables.py]:
```suggestion
        compute: MLPAlias | None = None,
```

`from causalprog._types import MLPAlias`

File: src/causalprog/graph/node/random_variables.py [hunk 15]
```diff
@@ -72,13 +73,14 @@ def sample(
     def evaluate(
         self,
         given_values: dict[str, jax.Array],
+        parameters: dict[str, jax.Array],
```
Comment (willGraham01) [inline src/causalprog/graph/node/random_variables.py]:
```suggestion
        parameters: ModelParam,
```

File: src/causalprog/graph/node/random_variables.py [hunk 16]
```diff
@@ -119,7 +128,8 @@ def __init__(
         values: list[float] | list[jax.Array],
         shape: tuple[int, ...] = (),
         label: str,
-        compute: Callable | None = None,
+        compute: Callable[[dict[str, jax.Array], dict[str, jax.Array]], jax.Array]
+        | None = None,
```
Comment (willGraham01) [inline src/causalprog/graph/node/random_variables.py]:
```suggestion
        compute: MLPAlias | None = None,
```

**Review by willGraham01 — CHANGES_REQUESTED** [review]
A few typehint suggestions given that we have `causalprog._types` available.

Other than that - the merge with `main` seems to have done a few funny things regarding function names in places. Though presumably the tests should pass once that's fixed?

File: docs/users/graph.md
Comment (mscroggs) [reply docs/users/graph.md:269]: `graph` here will still be as defined on lines 173-178. That graph takes parameters and this passes in parameters so I think we're ok?

**Review by willGraham01 — APPROVED** [review]
